[SPARK-14127][SQL] Native "DESC [EXTENDED | FORMATTED] <table>" DDL command

liancheng opened the pull request:
## What changes were proposed in this pull request?

This PR implements native `DESC [EXTENDED | FORMATTED] <table>` DDL command. Sample output:

```
scala> spark.sql("desc extended src").show(100, truncate = false)
+----------------------------+---------------------------------+-------+
|col_name                    |data_type                        |comment|
+----------------------------+---------------------------------+-------+
|key                         |int                              |       |
|value                       |string                           |       |
|                            |                                 |       |
|# Detailed Table Information|CatalogTable(`default`.`src`, ...|       |
+----------------------------+---------------------------------+-------+


scala> spark.sql("desc formatted src").show(100, truncate = false)
+----------------------------+----------------------------------------------------------+-------+
|col_name                    |data_type                                                 |comment|
+----------------------------+----------------------------------------------------------+-------+
|key                         |int                                                       |       |
|value                       |string                                                    |       |
|                            |                                                          |       |
|# Detailed Table Information|                                                          |       |
|Database:                   |default                                                   |       |
|Owner:                      |lian                                                      |       |
|Create Time:                |Mon Jan 04 17:06:00 CST 2016                              |       |
|Last Access Time:           |Thu Jan 01 08:00:00 CST 1970                              |       |
|Location:                   |hdfs://localhost:9000/user/hive/warehouse_hive121/src     |       |
|Table Type:                 |MANAGED                                                   |       |
|Table Parameters:           |                                                          |       |
|  transient_lastDdlTime     |1451898360                                                |       |
|                            |                                                          |       |
|# Storage Information       |                                                          |       |
|SerDe Library:              |org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe        |       |
|InputFormat:                |org.apache.hadoop.mapred.TextInputFormat                  |       |
|OutputFormat:               |org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat|       |
|Num Buckets:                |-1                                                        |       |
|Bucket Columns:             |[]                                                        |       |
|Sort Columns:               |[]                                                        |       |
|Storage Desc Parameters:    |                                                          |       |
|  serialization.format      |1                                                         |       |
+----------------------------+----------------------------------------------------------+-------+
```
## How was this patch tested?

A test case is added to `HiveDDLSuite` to check command output.

liancheng committed:
Implements native 'DESC TABE' DDL command

liancheng committed:
Test case

SparkQA commented:
**[Test build #57539 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57539/consoleFull)** for PR 12844 at commit [`718da25`](https://github.com/apache/spark/commit/718da25b489ef56e2b1092484bd6fbcdf2a547ed).

SparkQA commented:
**[Test build #57539 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57539/consoleFull)** for PR 12844 at commit [`718da25`](https://github.com/apache/spark/commit/718da25b489ef56e2b1092484bd6fbcdf2a547ed).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57539/
Test FAILed.

liancheng committed:
Fixes test failures

- Shows partition columns for EXTENDED and FORMATTED
- Shows "Compressed:" field
- Shows data types in lower case

SparkQA commented:
**[Test build #57597 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57597/consoleFull)** for PR 12844 at commit [`803f28e`](https://github.com/apache/spark/commit/803f28ed39c5013dfe63cfef404e78317cc1f980).

SparkQA commented:
**[Test build #57597 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57597/consoleFull)** for PR 12844 at commit [`803f28e`](https://github.com/apache/spark/commit/803f28ed39c5013dfe63cfef404e78317cc1f980).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57597/
Test FAILed.

SparkQA commented:
**[Test build #57601 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57601/consoleFull)** for PR 12844 at commit [`0bc9f5a`](https://github.com/apache/spark/commit/0bc9f5a7a604f2f10015ddef5049c692d6d97041).

SparkQA commented:
**[Test build #57601 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57601/consoleFull)** for PR 12844 at commit [`0bc9f5a`](https://github.com/apache/spark/commit/0bc9f5a7a604f2f10015ddef5049c692d6d97041).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57601/
Test FAILed.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -1012,7 +1017,7 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
       entry("field.delim", ctx.fieldsTerminatedBy) ++
         entry("serialization.format", ctx.fieldsTerminatedBy) ++
         entry("escape.delim", ctx.escapedBy) ++
-        entry("colelction.delim", ctx.collectionItemsTerminatedBy) ++
```
A typo bug, not related to this PR.

liancheng replied:
Oh wait... It's actually a typo from Hive... :dizzy_face:

SparkQA commented:
**[Test build #57613 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57613/consoleFull)** for PR 12844 at commit [`9194fe1`](https://github.com/apache/spark/commit/9194fe1112ed247ff10d37ee6864ef88a9bd07fb).

SparkQA commented:
**[Test build #57613 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57613/consoleFull)** for PR 12844 at commit [`9194fe1`](https://github.com/apache/spark/commit/9194fe1112ed247ff10d37ee6864ef88a9bd07fb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57613/
Test FAILed.

liancheng committed:
Reverts the "typo" fix

SparkQA commented:
**[Test build #57621 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57621/consoleFull)** for PR 12844 at commit [`a66885a`](https://github.com/apache/spark/commit/a66885a545d670aa5ddc94126284446c246ca81c).

SparkQA commented:
**[Test build #57621 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57621/consoleFull)** for PR 12844 at commit [`a66885a`](https://github.com/apache/spark/commit/a66885a545d670aa5ddc94126284446c246ca81c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57621/
Test FAILed.

liancheng committed:
Uses .simpleString instead of .sql when describing column data types

SparkQA commented:
**[Test build #57630 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57630/consoleFull)** for PR 12844 at commit [`18b9bb5`](https://github.com/apache/spark/commit/18b9bb50039d75150dca06b4f97bcffbe41343e7).

SparkQA commented:
**[Test build #57630 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57630/consoleFull)** for PR 12844 at commit [`18b9bb5`](https://github.com/apache/spark/commit/18b9bb50039d75150dca06b4f97bcffbe41343e7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57630/
Test PASSed.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:
```diff
@@ -48,6 +48,7 @@ case class CatalogStorageFormat(
     inputFormat: Option[String],
     outputFormat: Option[String],
     serde: Option[String],
+    compressed: Boolean,
```
Is this ever true? If it isn't we could leave it out.

hvanhovell replied:
Nvm. Hive can pass compressed tables.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala:
```diff
@@ -290,29 +291,88 @@ case class DescribeTableCommand(table: TableIdentifier, isExtended: Boolean)
+    buffer += Row("# Detailed Table Information", relation.catalogTable.toString, "")
+  }
+
+  private def describeFormatted(relation: CatalogRelation, buffer: ArrayBuffer[Row]): Unit = {
```
We could also use some sort of an `append(string, string, string)` function instead of calling `buffer += Row(..., ..., ...)` alot. But i guess that is just a matter of preference.

liancheng replied:
Makes sense, actually there were several times that I forgot to add the trailing empty string(s) while working on this PR. Thanks!

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala:
```diff
@@ -290,29 +291,88 @@ case class DescribeTableCommand(table: TableIdentifier, isExtended: Boolean)
+  private def describeSchema(schema: StructType, buffer: ArrayBuffer[Row]): Unit = {
+    schema.foreach { column =>
+      val comment =
+        if (column.metadata.contains("comment")) column.metadata.getString("comment") else ""
```
Just an idea: So 9/10 times I use `column.metadata` it is to get the comment. Shouldn't we just add this to the StructField...?

liancheng replied:
I think the metadata is heavily used in ML code. Another thing is that the data type API is already public for a long time. We probably don't want to change it unless there are particular good reasons.

hvanhovell replied:
I am not saying that we should (re)move it. I am only suggesting that it might be easier if we had such an accessor in StructField.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -1012,6 +1017,7 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
       entry("field.delim", ctx.fieldsTerminatedBy) ++
         entry("serialization.format", ctx.fieldsTerminatedBy) ++
         entry("escape.delim", ctx.escapedBy) ++
+        // The following typo is inherited from Hive...
```
+1 for this comment. Totally got me the last time.

hvanhovell commented:
LGTM

liancheng committed:
Addresses PR comment

SparkQA commented:
**[Test build #57729 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57729/consoleFull)** for PR 12844 at commit [`b5dbc15`](https://github.com/apache/spark/commit/b5dbc155afc447a0f700919f7645fc093ec26925).

SparkQA commented:
**[Test build #57729 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57729/consoleFull)** for PR 12844 at commit [`b5dbc15`](https://github.com/apache/spark/commit/b5dbc155afc447a0f700919f7645fc093ec26925).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57729/
Test PASSed.

liancheng commented:
Thanks for the review! I'm merging this to master and branch-2.0.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala:
```diff
@@ -290,29 +291,92 @@ case class DescribeTableCommand(table: TableIdentifier, isExtended: Boolean)
+    describe(relation, buffer)
+
+    append(buffer, "", "", "")
+    append(buffer, "# Detailed Table Information", relation.catalogTable.toString, "")
```
@liancheng To improve the output of `Explain`, I plan to change the default implementation of `toString` of `case class CatalogTable`. That will also affect the output of `Describe Extended`. 

I checked what Hive did for the command `Describe Extended`, as follows. 

`Detailed Table Information	Table(tableName:t1, dbName:default, owner:root, createTime:1462627092, lastAccessTime:0, retention:0, sd:StorageDescriptor(cols:[FieldSchema(name:col, type:int, comment:null)], location:hdfs://6b68a24121f4:9000/user/hive/warehouse/t1, inputFormat:org.apache.hadoop.mapred.TextInputFormat, outputFormat:org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat, compressed:false, numBuckets:-1, serdeInfo:SerDeInfo(name:null, serializationLib:org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, parameters:{serialization.format=1}), bucketCols:[], sortCols:[], parameters:{}, skewedInfo:SkewedInfo(skewedColNames:[], skewedColValues:[], skewedColValueLocationMaps:{}), storedAsSubDirectories:false), partitionKeys:[], parameters:{transient_lastDdlTime=1462627092}, viewOriginalText:null, viewExpandedText:null, tableType:MANAGED_TABLE)`

Basically, in the implementation of `toString`, I will try to follow what you did in `describeFormatted` but the contents will be in a single line. Feel free to let me know if you have any concern or suggestion. Thanks!
